duckduckgo/Android
# Set default browser experiment

Task/Issue URL: https://app.asana.com/0/1125189844152671/1132925309519722
Tech Design URL: https://app.asana.com/0/1125189844152671/1136673216829026
CC: 

**Description**:
Uses the system dialog to set DuckDuckGo as the default browser (if possible) rather than sending the user to the system settings.

**Steps to test this PR**:

**Pre steps**
1. In `OnboardingPageManager` change line 67 to `if (true) {` 
1. Make sure DDG is uninstalled 
1. Install a different browser and make it the default (you can’t use an already installed browser).

**Variants by manufacturer**
1. Install DDG in a Huawei device
1. Notice how the new variants for this experiment are filtered out when assigned

**Variants by api**
1. Install DDG in a device lower than Nougat
1. Notice how the new variants for this experiment are filtered out when assigned

**Variants by language**
1. Install DDG in a device where the default language is not English
1. Notice how the new variants for this experiment are filtered out when assigned

**UI tests**
Samsung S7 Edge

https://app.asana.com/0/inbox/1125189844075764/1139795300569647/1139802067653870
https://app.asana.com/0/inbox/1125189844075764/1139801011340353/1139802067653870

**Test click on Maybe Later**
1. Install DDG
1. Launch the app
1. Onboarding screen should show the dialog UI
1. Pixel sent `m_odb_v`
1. Click on maybe later
1. Should be taken to the browser
1. Pixel sent `m_odb_s`

**Test instructions are faded out but still shown if wait 3 seconds or more**
1. Install DDG
1. Launch the app
1. Onboarding screen should show the dialog UI
1. Pixel sent `m_odb_v`
1. Click on `Let’s Do It`
1. Pixel sent `m_odb_l` with `bt=d`
1. A dialog with the instructions card should be shown
1. Wait 3 seconds
1. Instructions card should fade out but still be visible.

**Test dismiss instructions and dialog**
1. Install DDG
1. Launch the app
1. Onboarding screen should show the dialog UI
1. Pixel sent `m_odb_v`
1. Click on `Let’s do it`
1. Pixel sent `m_odb_l` with `bt=d`
1. A dialog with the instructions card should be shown
1. Tap anywhere in the screen apart from the dialog or the back button
1. Dialog and instructions card should disappear
1. Pixel sent `m_db_ns` with `dbo=dd`
	
**Test select DDG and just once 2 times consecutively**
1. Install DDG
1. Launch the app
1. Onboarding screen should show the dialog UI
1. Pixel sent `m_odb_v`
1. Click on `Let’s do it`
1. Pixel sent `m_odb_l` with `bt=d`
1. A dialog with the instructions card should be shown
1. Select DDG and tap on `Just once`
1. Dialog and instructions card should be shown again
1. Pixel `m_obd_jo` sent
1. Tap on `Just once`
1. You should be taken to the browser and on boarding finished
1. Pixel sent `m_db_ns` with `dbo=jom`

**Test select DDG and Always**
1. Install DDG
1. Launch the app
1. Onboarding screen should show the dialog UI
1. Pixel sent `m_odb_v`
1. Click on `Let’s do it`
1. Pixel sent `m_odb_l` with `bt=d`
1. A dialog with the instructions card should be shown
1. Select DDG and tap on `Always`
1. Pixel sent `m_db_s` with `fo=true` and `dbo=d`
1. UI has changed confirming DDG is the default browser and showing a start browsing button
1. Tap start browsing
1. You should be taken to the browser and onboarding finished

**Test select a different browser using Just Once**
1. Install DDG
1. Launch the app
1. Onboarding screen should show the dialog UI
1. Pixel sent `m_odb_v`
1. Click on `Let’s do it`
1. Pixel sent `m_odb_l` with `bt=d`
1. A dialog with the instructions card should be shown
1. Select a different browser and tap just once (if available)
1. You should be taken to the selected browser and https://donttrack.us be loaded
1. Press the back button
1. You should be taken back to the onboarding screen
1. Pixel sent `m_db_ns` with `dbo=e`

At this point, in some android versions, once you tap on `Let’s Do It` you might have the selected browser as the pre-selected option in the dialog and DDG will be under the “Use a different app” section. If this is case then:
1. Select DDG
1. Dialog should be shown again but DDG is now the pre-selected option

**Test a different browser was already selected as the default**
1. Install DDG and do not launch it
1. Install a different browser and make it the default browser
1. Launch DDG
1. Onboarding screen should show the settings UI
1. Pixel sent `m_odb_v`
1. Tap on `Let’s do it` and should be taken to the settings
1. Pixel sent `m_odb_l` with `bt=s`
1. Select DDG and press back
1. Pixel sent `m_db_s` with `fo=true` and `dbo=s`
1. UI has changed confirming DDG is the default browser and showing a start browsing button
1. Navigate back to the first onboarding screen and notice how the UI still shows DDG is the default browser
1. Close the app and relaunch again and notice how the UI still shows DDG is the default browser
1. Tap start browsing
1. You should be taken to the browser and onboarding finished

**Test user selects a different browser as their default with Always**
1. Install DDG and another browser
1. Launch DDG 
1. Pixel sent `m_odb_v`
1. Tap on `Let’s do it`
1. Pixel sent `m_odb_l` with `bt=d`
1. Select other browser using `Always`
1. Press back and you should now see the settings UI in the onboarding screen
1. Pixel sent `m_db_ns` with `dbo=e`

**Test user re-installing DDG when they had a default browser selected along with DDG in the past**
1. Install DDG and another browser
1. Launch DDG 
1. Pixel sent `m_odb_v`
1. Tap on `Let’s do it`
1. Select other browser using `Always`
1. Uninstall DDG
1. Install DDG again
1. The browser select as always takes precedence hence you should see the settings UI.

**Test pixel with correct parameter when setting DDG as the default browser externally**
1. Install DDG
1. Launch DDG and skip onboarding
1. Close DDG
1. Set DDG as the default browser using the android settings
1. Launch DDG
1. Pixel sent `m_db_s` with `fo=false`

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
